pkg/resource/node_resource_manager.go

```diff
@@ -0,0 +1,232 @@
+		}
+		return
+	}
+}
+
+func (o *NodeResourceManager) getNode() *v1.Node {
+	node, err := o.nodeLister.Get(o.nodeName)
+	if err != nil {
+		klog.Errorf("Failed to get node: %v", err)
+		return nil
+	}
+	return node
+}
+
+func (o *NodeResourceManager) FindTargetNode(tsp *predictionapi.TimeSeriesPrediction, addresses []v1.NodeAddress) (bool, error) {
+	address := tsp.Spec.TargetRef.Name
+	if address == "" {
+		return false, fmt.Errorf("Tsp %s target is not specified", tsp.Name)
+	}
+
+	// the reason we use node ip instead of node name as the target name is
+	// some monitoring system does not persist node name
+	for _, addr := range addresses {
+		if addr.Address == address {
+			return true, nil
+		}
+	}
+	klog.V(4).Infof("Target %s mismatch this node", address)
+	return false, nil
+}
+
+func (o *NodeResourceManager) BuildNodeStatus(tsp *predictionapi.TimeSeriesPrediction, node *v1.Node) {
+	idToResourceMap := map[string]*v1.ResourceName{}
+	for _, metrics := range tsp.Spec.PredictionMetrics {
+		if metrics.ResourceQuery == nil {
+			continue
+		}
+		idToResourceMap[metrics.ResourceIdentifier] = metrics.ResourceQuery
+	}
+	// build node status
+	nextPredictionResourceStatus := &tsp.Status
+	for _, metrics := range nextPredictionResourceStatus.PredictionMetrics {
+		resourceName, exists := idToResourceMap[metrics.ResourceIdentifier]
+		if !exists {
+			continue
+		}
+		for _, timeSeries := range metrics.Prediction {
+			var maxUsage, nextUsage float64
+			var nextUsageFloat float64
+			var err error
+			for _, sample := range timeSeries.Samples {
+				if nextUsageFloat, err = strconv.ParseFloat(sample.Value, 64); err != nil {
+					klog.Errorf("Failed to parse extend resource value %v: %v", sample.Value, err)
+					continue
+				}
+				nextUsage = nextUsageFloat
+				if maxUsage < nextUsage {
+					maxUsage = nextUsage
+				}
+			}
+			var nextRecommendation float64
+			switch *resourceName {
+			case v1.ResourceCPU:
+				// cpu need to be scaled to m as ext resource cannot be decimal
+				nextRecommendation = (float64(node.Status.Allocatable.Cpu().Value()) - maxUsage) * 1000
+			case v1.ResourceMemory:
+				// unit of memory in prometheus is in Ki, need to be converted to byte
+				nextRecommendation = float64(node.Status.Allocatable.Memory().Value()) - (maxUsage * 1000)
+			default:
+				continue
+			}
+			if nextRecommendation < 0 {
+				klog.V(4).Infof("Unexpected recommendation,nodeName %s, maxUsage %v, nextRecommendation %v", node.Name, maxUsage, nextRecommendation)
+				continue
+			}
+			extResourceName := fmt.Sprintf(executor.ExtResourcePrefixFormat, string(*resourceName))
+			resValue, exists := node.Status.Capacity[v1.ResourceName(extResourceName)]
+			if exists && resValue.Value() != 0 &&
+				math.Abs(float64(resValue.Value())-
+					nextRecommendation)/float64(resValue.Value()) <= MinDeltaRatio {
+				continue
+			}
+			node.Status.Capacity[v1.ResourceName(extResourceName)] =
```

Comment:
if resource is cpu ，should we use binarySI here？

Reply:
cpu should use DecimalSI in my opinion?

Reply:
ok, what if the resource is memory?

Reply:
BinarySI